### PR TITLE
refactor: enforce a cognitive complexity limit and clear the 28 violations

### PR DESCRIPTION
**Pull request by AbianS**
## What this is

`biome.json` turns on `noExcessiveCognitiveComplexity` at the default ceiling of
15. That surfaced **28 violations** across `apps/webview-ui` and `packages/ui`,
and this PR clears all of them.

**No `biome-ignore` anywhere.** A complexity rule that gets waived in place is
worse than no rule, because it reads as enforced while the exceptions accumulate.

One framing point that shaped the whole approach: Biome measures *cognitive*
complexity, not cyclomatic. It weights **nesting**, so `if` inside `if` inside a
`.map()` costs 1+2+3 while three sibling `if`s cost 1+1+1. Flattening therefore
pays far more than deleting branches, and every refactor below moves work into
named units rather than compressing it.

## The 28 were six shapes, not 28 problems

| Pattern | Count | Worst | Fix |
|---|---|---|---|
| Inline `.map()` row renderers | 7 | 27 | Named row components + shared `barGeometry` |
| Keyboard handlers as if-chains | 4 | **35** | Key-to-intent dispatch |
| Provider-type validation chains | 3 | 25 | `Record<ProviderType, ...>` strategy table |
| Triplicated form `onSubmit` | 3 | 21 | One `useIntegrationSubmit` hook |
| Server page: load + derive + render | 3 | 25 | Pure, tested derivation functions |
| Query codecs and fixtures | 5 | **61** | `FilterCodec` / predicate tables |

### The two that mattered most

**`fakeServer` at 61** (`data-table.stories.tsx`) was an if-chain per column
inside a `.filter()`. It is now a `MATCHERS` predicate table. Worth doing even in
a story file, because that fake server is the de facto documentation of the
filter behaviour the Rust endpoint is expected to match, and at 61 nobody was
reading it.

**`onKeyDown` at 35** (`use-query-bar.ts`) was six sequential
`if (event.key === ...)` blocks with nested branches inside each. It is now a
flat `switch` delegating to five named intents (`moveHighlight`, `commitOrApply`,
`completeField`, `dismiss`, `stepBack`), none above 5.

### Type safety gained on the way

Three of the patterns replace an if-chain with a keyed table, and the key type is
the point rather than a side effect:

- `Record<ProviderType, RoutingBehaviour>` in `routing.ts`
- `Record<ColumnFilterSpec['variant'], FilterCodec>` in `query.ts`
- `Record<ColumnFilterSpec['variant'], Panel>` in `column-header.tsx`

Adding a provider or a filter variant is now a **compile error** until it is
handled everywhere. The if-chains it replaces silently fell through, which is how
`routing.ts` ended up with the same `provider_type` chain written three times in
three functions, where adding a provider was three edits and nothing failed if
only two of them happened.

## Behaviour is unchanged, and that is tested

Every structural refactor landed either test-first or behind characterization
tests written against the existing code **before** it was touched. That is what
the **93 new unit tests** are:

`routing` (29), `integration-forms` (13), `trace-summary` (12), `transaction-payload` (8),
`waterfall-geometry` (8), `detectPhase` (7), `credentials` (6), `event-jumps` (4),
`setup-snippet` (4), plus 2 pinning existing `query.ts` edge cases.

- `packages/ui`: 92 to 185 tests
- `apps/webview-ui`: 92 to 170 tests
- `pnpm run ci`: 27/27 green, Rust suite included

## One regression, caught by the Storybook tests

Extracting the column header's trigger `<button>` into its own component broke
three Storybook interaction tests. Base UI's `Popover.Trigger` takes the element
through a `render` prop and **merges** the click handler, the ref and
`data-popup-open` into it. A function component received all of that as props it
did not forward, leaving a button that never opened its popover.

Neither TypeScript nor the linter would have caught it. The `<button>` element is
back inline with a comment explaining why it cannot move, and only its children
are extracted. Worth flagging as an argument for keeping those play tests.

## Two pre-existing issues found

**`buildChannelsPayload` was quadratic.** It searched with `integrations.find()`
inside a `flatMap` over the selection, directly below `collectRoutingErrors`,
which had already indexed the same pair with a `Map` and carried a comment
explaining why. It uses the `Map` now.

**An undocumented asymmetry in the query parser.** `level:,,` (a malformed option
list) is silently dropped, while `events:oops` (a malformed range) is kept as free
text. Nothing tested it and nothing said it was deliberate. Both are now pinned by
tests and the third outcome is named `drop` in the codec type, so the difference
is stated rather than implied. **The behaviour itself is unchanged**, by request.

## Also in here

**`vitest.config.mts` gains a `@/` alias.** Type-only `@/` imports are erased at
compile time, so its absence never showed until a test's subject imported a real
*value* through one, which fails as "Cannot find package" rather than as missing
config. That file is deliberately minimal and heavily commented, so the addition
says exactly which gap it closes.

**A separate commit fixes CI on non-`main` PRs.** `branches:` under a
`pull_request` trigger filters on the PR's *base*, so gating on `main` meant a PR
onto a feature branch ran no CI, no CodeQL and no cargo-deny at all. CodeRabbit
had the same problem from its side: it only reviews PRs onto the default branch
unless given `base_branches`. Both are fixed.

> Note: branch protection is configured in GitHub, not in the repo. These checks
> now *run* on such PRs; whether they *block* a merge is a separate setting worth
> checking.

## Reviewing this

The diff is large but repetitive: seven row extractions and three table
conversions account for most of it. The parts worth real attention are
`routing.ts`, `query.ts` and `use-integration-submit.ts`, where the shape of the
code actually changed rather than just its location.

## Summary by CodeRabbit

* **New Features**
  * Improved event detail pages with sticky headers, navigation, jump links, timestamps, release information, and activity summaries.
  * Added clearer agent trace summaries, span selection states, detail panes, and waterfall visualizations.
  * Improved transaction detail badges and project client-key setup instructions.
  * Added reliable copy feedback, including clipboard-unavailable notifications.

* **Bug Fixes**
  * Improved alert integration submissions, validation, credential handling, and error display.
  * Improved query filtering for empty options and malformed ranges.
  * Improved waterfall bar positioning and keyboard interactions.

### CODE REVIEW

File: apps/webview-ui/src/app/(main)/projects/[id]/performance/[txnId]/_components/transaction-badges.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import type { TransactionDetail } from '@rustrak/client';
+import { getFormatter } from 'next-intl/server';
+import { Badge } from '@/shared/ui/components/shadcn/badge';
+
+function formatDuration(ms: number | null): string {
+  if (ms === null) return '—';
+  if (ms < 1000) return `${Math.round(ms)}ms`;
+  return `${(ms / 1000).toFixed(2)}s`;
```
[the remark this answers is not in the document]
Comment (AbianS) [reply apps/webview-ui/src/app/(main)/projects/[id]/performance/[txnId]/_components/transaction-badges.tsx:8]: > Format displayed durations with the repository's locale-aware formatter.

Reviewed: not changed in this PR. `formatDuration` is byte-identical to `HEAD~2` in all three sites; this PR only relocated it while extracting components, so there is no regression here to fix.

The underlying point is fair though. `CLAUDE.md` says a number is formatted at the call site, and these badges already use `format.number` and `format.dateTime` for everything else, so the duration is the odd one out across five locales. It is a visible output change rather than a refactor cleanup, so it belongs in its own PR where the wording per locale can actually be looked at.

File: apps/webview-ui/src/features/transaction/lib/transaction-payload.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ */
+export function readTransactionPayload(
+  txn: TransactionDetail,
+): TransactionPayload {
+  const data = (txn.data ?? {}) as Record<string, unknown>;
+  const trace = asObject(asObject(data.contexts)?.trace) ?? undefined;
+
+  return {
+    trace: trace as TraceContext | undefined,
+    spans: Array.isArray(data.spans) ? (data.spans as Span[]) : [],
```
[the remark this answers is not in the document]
Comment (AbianS) [reply apps/webview-ui/src/features/transaction/lib/transaction-payload.ts:60]: > Validate nested trace fields and every span element.

Reviewed: not changed in this PR. `readTransactionPayload` is an extraction of code that already lived in the page, and the guarantees are exactly what they were: `Array.isArray(data.spans) ? data.spans : []` followed by a cast, identical to `HEAD~2`. No new risk was introduced.

Whether the UI should defensively parse every element of an SDK-supplied payload is a real question, but it is a design decision about the whole `data` blob rather than something to bolt onto a complexity refactor. If it gets taken on, the right shape is a Zod schema at the boundary, applied consistently, not per-field guards in this one function.

File: apps/webview-ui/src/shared/ui/components/data-table/data-table.tsx [hunk 3]
```diff
@@ -274,87 +274,118 @@ export function DataTable<TData extends RowData>({
...
+  row: Row<DataTableFeatures, TData>;
+  /** Held only for `FlexRender`, which is a method on the instance. */
+  table: DataTableInstance<TData>;
+  spacing: (typeof DENSITY)[keyof typeof DENSITY];
+  /** Width of the trailing filler cell, or 0 when the columns fill the table. */
+  fillerWidth: number;
+  onRowClick?: (row: Row<DataTableFeatures, TData>) => void;
+  getRowClassName?: (row: Row<DataTableFeatures, TData>) => string | undefined;
+  renderDetail?: (row: Row<DataTableFeatures, TData>) => ReactNode;
+}
```
[the remark this answers is not in the document]
Comment (AbianS) [reply apps/webview-ui/src/shared/ui/components/data-table/data-table.tsx:307]: > Keep table-model props out of this shared UI component.

Reviewed: the rule is real but it is being applied to the wrong thing.

`Props are only primitives or ReactNode` comes from the **"Where does this file go?"** table in `apps/webview-ui/CLAUDE.md`. That table decides which folder a component belongs in, by the type its props name. It is a placement heuristic, not a prop-purity constraint on code already placed.

`DataTableRow` is a private, non-exported helper of `DataTable` in the same file, and it takes a strict subset of what the exported `DataTable` already takes: `table: DataTableInstance<TData>`, `renderDetail?: (row) => ReactNode`, `onRowClick?: (row) => void`, `getRowClassName?: (row) => string | undefined`. If the rule meant what is suggested here, it would condemn the pre-existing public component more than the private helper extracted from it.

The proposed move is also not available: `shared/` importing from `features/` inverts the layer direction, which `src/__tests__/architecture/layer-direction.test.ts` fails on outright. And the alternative offered, "a local render helper instead of a component", is what it was before this PR, which is precisely what put the function over the complexity ceiling.

No change.

**Review by AbianS — COMMENTED** [review]
All 8 findings analyzed against the code. **5 fixed in 98c86c4**, 3 answered as false positives in their threads.

Fixed:

- `isHttpUrl` prefix-matched, so `https://` and `https:// invalid` passed routing validation. Now parsed with `URL`, protocol checked, four malformed values pinned by tests.
- The span waterfall's collapse button stopped click propagation but not keydown, and the row handler calls `preventDefault`, which is what dispatches a native button's click. Enter or Space on the chevron selected the row and never expanded it. This was the best catch in the review.
- `barGeometry` returned a negative width for a span starting after the window. Both ends clamped, with the suggested test case plus three more.
- Column header gained `pressScaleTrigger`, per the Base UI note in `packages/ui/CLAUDE.md`.
- Toast progress normalized once, so `aria-valuenow` and the caption can no longer report 150 against a declared maximum of 100.

Not changed, with reasoning in each thread: the shared-UI prop rule cited is a file-placement heuristic rather than a prop ban, and the move it proposes would invert the layer direction the architecture tests enforce; the other two flag code this PR only relocated, byte-identical to `HEAD~2`.

Verification: `packages/ui` 185 tests, `webview-ui` 177 tests, `tsc` and Biome clean, `next build` succeeds.
